src/compiled_proof.rs

```diff
@@ -0,0 +1,437 @@
+            data,
+            t: PhantomData,
+        }
+    }
+}
+
+impl<'a, T: Packable> Iterator for PackedLeaves<'a, T> {
+    type Item = Result<(u64, T)>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index >= self.data.len() {
+            return None;
+        }
+
+        if self.data.len() - self.index < 8 {
+            return Some(Err(Error::UnpackEof));
+        }
+
+        let pos = {
+            let mut buf = [0u8; 8];
+            buf.copy_from_slice(&self.data[self.index..self.index + 8]);
+            u64::from_le_bytes(buf)
+        };
+        self.index += 8;
+
+        match T::unpack(&self.data[self.index..]) {
+            Ok((item, size)) => {
+                self.index += size;
+                Some(Ok((pos, item)))
+            }
+            Err(e) => Some(Err(e)),
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+enum StackItemType {
+    Node,
+    Peak,
+    Proof,
+}
+
+pub fn calculate_root<
+    T: PartialEq + Clone,
+    M: Merge<Item = T>,
+    I: Iterator<Item = Result<Command<T>>>,
+    IT: Iterator<Item = Result<(u64, T)>>,
+>(
+    commands: &mut I,
+    mmr_size: u64,
+    leaves: &mut IT,
+) -> Result<T> {
+    let mut last_leaf_pos = None;
+
+    let mut next_peak_info = PeakIterator::new(mmr_size);
+    // let mut leaf_index = 0;
+    let mut stack: Vec<(T, u64, u8, StackItemType)> = Vec::with_capacity(257);
+    for command in commands {
+        match command? {
+            Command::NextLeaf => {
+                let (pos, leaf_item) = leaves.next().ok_or(Error::CorruptedStack)??;
+                if pos_height_in_tree(pos) > 0 {
+                    return Err(Error::NodeProofsNotSupported);
+                }
+                if let Some(last_leaf_pos) = last_leaf_pos {
+                    // ensure leaves are sorted and unique
+                    if last_leaf_pos >= pos {
+                        return Err(Error::CorruptedProof);
+                    }
+                }
+                last_leaf_pos = Some(pos);
+                stack.push((leaf_item, pos, 0, StackItemType::Node));
+            }
+            Command::Proof(proof) => {
+                stack.push((proof.clone(), 0, 0, StackItemType::Proof));
+            }
+            Command::Hash => {
+                if stack.len() < 2 {
+                    return Err(Error::CorruptedStack);
+                }
+                let (rhs, rhs_pos, rhs_height, rhs_t) = stack.pop().unwrap();
+                let (lhs, lhs_pos, lhs_height, lhs_t) = stack.pop().unwrap();
+                let (pos, height, next_pos, next_t, item, sibling_item) =
+                    if lhs_t == StackItemType::Proof {
+                        // Swap lhs with rhs in case lhs is a proof
+                        (rhs_pos, rhs_height, lhs_pos, lhs_t, rhs, lhs)
+                    } else {
+                        (lhs_pos, lhs_height, rhs_pos, rhs_t, lhs, rhs)
+                    };
+                // calculate sibling
```

Comment:
I think the height var in stack is only used in `Command::Hash`, can we remove it from the stack and calculate it here?

```suggestion
                // calculate sibling
                let height = pos_height_in_tree(pos);
```

Reply:
I tried this, but noticing that recalculating `pos_height_in_tree` would have a noticeable impact on verifier performance (~2% in cycle consumptions of Rust code), so I would rather suggest keeping the code as it is. Counting heights only by incrementing as in current code, can be faster.

Reply:
> I tried this, but noticing that recalculating `pos_height_in_tree` would have a noticeable impact on verifier performance (~2% in cycle consumptions of Rust code), so I would rather suggest keeping the code as it is. Counting heights only by incrementing as in current code, can be faster.

I see, can you help to try the cycle difference between `StackItemType::Node(u64, u8) // pos, height` and original solution?

Reply:
This additional commit https://github.com/nervosnetwork/merkle-mountain-range/pull/27/commits/b5b4602cabe060101025c8df9e791167b5fb78ef which refactors as you mentioned, takes an additional 2000 cycles out of a total of 1583330 cycles, which is quite neglegible, hence I pull the trigger and include the refactored code.